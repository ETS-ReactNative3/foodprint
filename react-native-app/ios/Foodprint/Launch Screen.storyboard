<?xml version="1.0" encoding="UTF-8"?>
<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="16096" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" launchScreen="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="01J-lp-oVM">
    <device id="ipad10_2" orientation="portrait" layout="fullscreen" appearance="light"/>
    <dependencies>
        <deployment identifier="iOS"/>
        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="16087"/>
        <capability name="Safe area layout guides" minToolsVersion="9.0"/>
        <capability name="documents saved in the Xcode 8 format" minToolsVersion="8.0"/>
    </dependencies>
    <scenes>
        <!--View Controller-->
        <scene sceneID="EHf-IW-A2E">
            <objects>
                <viewController id="01J-lp-oVM" sceneMemberID="viewController">
                    <view key="view" contentMode="scaleToFill" id="Ze5-6b-2t3">
                        <rect key="frame" x="0.0" y="0.0" width="810" height="1080"/>
                        <autoresizingMask key="autoresizingMask" widthSizable="YES" heightSizable="YES"/>
                        <subviews>
                            <imageView clipsSubviews="YES" userInteractionEnabled="NO" contentMode="scaleAspectFit" horizontalHuggingPriority="251" verticalHuggingPriority="251" fixedFrame="YES" image="WhiteLogo" translatesAutoresizingMaskIntoConstraints="NO" id="Gol-wY-bC1">
                                <rect key="frame" x="270" y="425" width="271" height="231"/>
                                <autoresizingMask key="autoresizingMask" flexibleMinX="YES" widthSizable="YES" flexibleMaxX="YES" flexibleMinY="YES" heightSizable="YES" flexibleMaxY="YES"/>
                            </imageView>
                        </subviews>
                        <color key="backgroundColor" red="0.36078431372549019" green="0.36078431372549019" blue="0.36078431372549019" alpha="1" colorSpace="custom" customColorSpace="sRGB"/>
                        <viewLayoutGuide key="safeArea" id="Bcu-3y-fUS"/>
                    </view>
                </viewController>
                <placeholder placeholderIdentifier="IBFirstResponder" id="iYj-Kq-Ea1" userLabel="First Responder" sceneMemberID="firstResponder"/>
            </objects>
            <point key="canvasLocation" x="51.851851851851848" y="374.44444444444446"/>
        </scene>
    </scenes>
    <resources>
        <image name="WhiteLogo" width="512" height="512"/>
    </resources>
</document>
