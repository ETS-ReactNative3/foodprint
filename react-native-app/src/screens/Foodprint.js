import React, {useEffect, useState} from 'react';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import {
  View,
  ScrollView,
  StyleSheet,
  SafeAreaView,
  Text, ActivityIndicator,
} from 'react-native';
import { Button } from 'react-native-elements';
import { widthPercentageToDP as percentageWidth, heightPercentageToDP as percentageHeight } from 'react-native-responsive-screen';
import { TouchableOpacity } from 'react-native-gesture-handler';
import WelcomeScreen from '../components/WelcomeScreen';
import WeeklyDisplay from '../components/WeeklyDisplay';
import MonthlyDisplay from '../components/MonthlyDisplay';
import GeneralDisplay from '../components/GeneralDisplay';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/react-hooks';
import AsyncStorage from '@react-native-community/async-storage';


//DO NOT DELETE THE FOLLOWING COMMENTED CODE
export const GET_WEEKLY_AVERAGE = gql`query($timezone: Int!) {
  getPeriodAvg(timezone: $timezone, resolution: WEEK)
}`;

//DO NOT DELETE THE FOLLOWING COMMENTED CODE
export const GET_WEEKLY_COMPOSITION = gql`query($timezone: Int!) {
  reportByCategory(timezone: $timezone, resolution: WEEK) {
     plantBased {
      periodNumber
      avgCarbonFootprint
    }
     fish {
      periodNumber
      avgCarbonFootprint
    }
    meat {
      periodNumber
      avgCarbonFootprint
    }
    eggsAndDairy {
      periodNumber
      avgCarbonFootprint
    }
   }
 }`;

//DO NOT DELETE THE FOLLOWING COMMENTED CODE
export const GET_MONTHLY_AVERAGE = gql`query($timezone: Int!) {
  getPeriodAvg(timezone: $timezone, resolution:MONTH)
}`;

//DO NOT DELETE THE FOLLOWING COMMENTED CODE
export const GET_MONTHLY_COMPOSITION = gql`query($timezone: Int!) {
  reportByCategory(timezone: $timezone, resolution:MONTH) {
     plantBased {
      periodNumber
      avgCarbonFootprint
    }
     fish {
      periodNumber
      avgCarbonFootprint
    }
    meat {
      periodNumber
      avgCarbonFootprint
    }
    eggsAndDairy {
      periodNumber
      avgCarbonFootprint
    }
   }
 }`;


const Foodprint = ({ navigation }) => {

  const [isVisible, setVisibility] = useState(true);
  const [timeSpan, setTimeSpan] = useState('weekly');
  const [weeklyResolved, setWeeklyResolved] = useState(false);
  const [localWeeklyAvg, setWeeklyAvg] = useState(null);
  const [localWeeklyComp, setWeeklyComp] = useState(null);
  const [localMonthlyAvg, setMonthlyAvg] = useState(null);
  const [localMonthlyComp, setMonthlyComp] = useState(null);
  const [monthlyResolved, setMonthlyResolved] = useState(false);

  const takePicture = async () => {
    navigation.navigate('Camera');
  };

  const getTimeDifference = () => {
    const date = new Date();
    const timeDifference = date.getTimezoneOffset();
    return (timeDifference / 60);
  };

  // Comment the following lines to test the caching
  const { loading: weeklyAvgLoading, error: weeklyAvgError, data: weeklyAvgData } = useQuery(GET_WEEKLY_AVERAGE, {
    variables: { timezone: getTimeDifference() },
  });
  const { loading: weeklyCompLoading, error: weeklyCompError, data: weeklyCompData } = useQuery(GET_WEEKLY_COMPOSITION, {
    variables: { timezone: getTimeDifference() },
  });

  // Uncomment the following lines to test the caching
  // let weeklyAvgLoading = false;
  // let weeklyCompLoading = false;
  // let weeklyAvgError = true;
  // let weeklyCompError = true;
  // let weeklyAvgData = null;
  // let weeklyCompData = null;

  // Comment the following lines to test the caching
  const { loading: monthlyAvgLoading, error: monthlyAvgError, data: monthlyAvgData } = useQuery(GET_MONTHLY_AVERAGE, {
    variables: { timezone: getTimeDifference() },
  });
  const { loading: monthlyCompLoading, error: monthlyCompError, data: monthlyCompData } = useQuery(GET_MONTHLY_COMPOSITION, {
    variables: { timezone: getTimeDifference() },
  });

  // Uncomment the following lines to test the caching
  // let monthlyAvgLoading = false;
  // let monthlyCompLoading = false;
  // let monthlyAvgError = true;
  // let monthlyCompError = true;
  // let monthlyAvgData = null;
  // let monthlyCompData = null;


  const saveWeeklyData = async (weeklyAverage, weeklyComposition) => {
    try {
      await AsyncStorage.setItem('weeklyAverage', weeklyAverage);
      await AsyncStorage.setItem('weeklyComposition', weeklyComposition);
    } catch (e) {
      console.log('Error saving weekly data' + e);
    }
  };

  const saveMonthlyData = async (monthlyAverage, monthlyComposition) => {
    try {
      await AsyncStorage.setItem('monthlyAverage', monthlyAverage);
      await AsyncStorage.setItem('monthlyComposition', monthlyComposition);
    } catch (e) {
      console.log('Error saving monthly data' + e);
    }
  };


  const retrieveWeeklyData = async () => {
    try {
      const retrievedAvgData = await AsyncStorage.getItem('weeklyAverage');
      setWeeklyAvg(JSON.parse(retrievedAvgData));
      const retrievedCompData = await AsyncStorage.getItem('weeklyComposition');
      setWeeklyComp(JSON.parse(retrievedCompData));
    } catch (e) {
      console.log('Error retrieving weekly data ' + e);
    }
    setWeeklyResolved(true);
  };

  const retrieveMonthlyData = async () => {
    try {
      const retrievedAvgData = await AsyncStorage.getItem('monthlyAverage');
      setMonthlyAvg(JSON.parse(retrievedAvgData));
      const retrievedCompData = await AsyncStorage.getItem('monthlyComposition');
      setMonthlyComp(JSON.parse(retrievedCompData));
    } catch (e) {
      console.log('Error retrieving monthly data ' + e);
    }
    setMonthlyResolved(true);
  };


  useEffect(() => {
    if (weeklyAvgData && weeklyCompData) {
      saveWeeklyData(JSON.stringify(weeklyAvgData),JSON.stringify(weeklyCompData));
    }
  });

  useEffect(() => {
    if (monthlyAvgData && monthlyCompData) {
      saveMonthlyData(JSON.stringify(monthlyAvgData),JSON.stringify(monthlyCompData));
    }
  });


  useEffect(() => {
    if (weeklyAvgError || weeklyCompError) {
      retrieveWeeklyData();
    }
    if (monthlyAvgError || monthlyCompError) {
      retrieveMonthlyData();
    }
  });

  const whichWeeklyAvg = () => {
    if (weeklyAvgData) {return weeklyAvgData;}
    return localWeeklyAvg;
  };

  const whichWeeklyComp = () => {
    if (weeklyCompData) {return weeklyCompData;}
    return localWeeklyComp;
  };

  const whichMonthlyAvg = () => {
    if (monthlyAvgData) {return monthlyAvgData;}
    return localMonthlyAvg;
  };

  const whichMonthlyComp = () => {
    if (monthlyCompData) {return monthlyCompData;}
    return localMonthlyComp;
  };

  return (
    <SafeAreaView>
      <ScrollView>

        {/*General carbon footprint score*/}
        <View>
          <GeneralDisplay/>
        </View>

        {/*Monthly vs Weekly buttons*/}
        <View style={ styles.buttonContainer }>
          <Button
            title="Weekly"
            titleStyle={ styles.buttonTitle }
            buttonStyle={ [ styles.button, {backgroundColor: ((timeSpan === 'weekly') ? 'green' : 'grey')} ] }
            containerStyle={{ paddingHorizontal: percentageWidth('2%') }}
            onPress={() => setTimeSpan('weekly')}
          />
          <Button
            title="Monthly"
            titleStyle={ styles.buttonTitle }
            buttonStyle={ [ styles.button, {backgroundColor: ((timeSpan === 'monthly') ? 'green' : 'grey')}] }
            containerStyle={{ paddingHorizontal: percentageWidth('2%') }}
            onPress={() => setTimeSpan('monthly')}
          />
        </View>

        {/*Monthly or weekly carbon footprint composition display*/}
        <View>
          {(timeSpan === 'weekly') ? (
            (weeklyAvgLoading || weeklyCompLoading || ((weeklyAvgError || weeklyCompError) && !weeklyResolved)) ? (
              <View style={ styles.graphContainer }>
                <ActivityIndicator/>
              </View>
            ) : (
              <WeeklyDisplay average={whichWeeklyAvg()} composition={whichWeeklyComp()}/>
              )
          ) : (
            (monthlyAvgLoading || monthlyCompLoading || ((monthlyAvgError || monthlyCompError) && !monthlyResolved)) ? (
              <View style={ styles.graphContainer }>
                <ActivityIndicator/>
              </View>
            ) : (
                <MonthlyDisplay average={whichMonthlyAvg()} composition={whichMonthlyComp()}/>
                )
            )}
        </View>
        <View style={ styles.footnote }>
          <Text style={{ fontSize:percentageWidth('3%')}}>The carbon footprint displayed in this app,
            including this page, are given in kilograms of CO2 per kilogram of food. The weight of any food item is
            systematically normalised to 1kg to get to this result.</Text>
        </View>
        <WelcomeScreen setVisibility={setVisibility} isVisible={isVisible}/>
      </ScrollView>

      {/*Camera button*/}
      <TouchableOpacity onPress={takePicture} containerStyle={ styles.camera }>
        <MaterialCommunityIcons name="camera" color={'white'} size={28} />
      </TouchableOpacity>
      {/* <FAB buttonColor="#008000" iconTextColor="#FFFFFF" onClickAction={takePicture} visible={true} iconTextComponent={} /> */}
    </SafeAreaView >
  );
};

const styles = StyleSheet.create({
  score: { fontSize: percentageWidth('6%'), color: 'grey' },
  footnote: { margin:percentageWidth('5%'), marginTop:percentageHeight('7%'), marginBottom:percentageHeight('15%') },
  buttonContainer: { flexDirection: 'row', justifyContent: 'center', paddingVertical: percentageHeight('2%') },
  buttonTitle: { fontSize: percentageWidth('5%') },
  button: { width: percentageWidth('30%'), height: 45 },
  camera: { backgroundColor: '#008000', width: 64, height: 64, position: 'absolute', bottom: 25, right: 25, borderRadius: 32, alignItems: 'center', justifyContent: 'center' },
  graphContainer: { height: percentageHeight('38%'), justifyContent: 'center', marginVertical: percentageHeight('2%') },
});

export default Foodprint;
