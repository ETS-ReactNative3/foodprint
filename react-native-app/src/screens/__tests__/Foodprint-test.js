import React from 'react';

import renderer from 'react-test-renderer';
import Foodprint from '../Foodprint';
import { MockedProvider } from '@apollo/react-testing';
import { mockedResponses } from '../../__tests__/.mockedResponses';

// Check current component against its snapshot
test('foodprint screen with loaded results matches previous snapshot', async () => {
  const mockedNavigation = {
    setOptions: jest.fn(),
  };

  let component;
  renderer.act(() => {
    component = <MockedProvider mocks={mockedResponses} addTypename={false}>
      <Foodprint navigation={mockedNavigation} />
    </MockedProvider>;
  });

  expect(component).toMatchSnapshot();
});

// test('foodprint screen caches all responses', async () => {
//   const mockedNavigation = {
//     setOptions: jest.fn(),
//   };

//   let component;
//   renderer.act(() => {
//     component = <MockedProvider mocks={mocks} addTypename={false}>
//       <Foodprint navigation={mockedNavigation} />
//     </MockedProvider>;
//   });

//   await wait(0);

//   // UI's are hard to test and we currently cannot ensure the useEffect hooks are triggered in a test.
//   // They are usually triggered based on state change events, but we cannot change the state directly in a test.
//   // Instead, we mock the response from the network, which currently does not trigger useEffect hooks.
//   expect(AsyncStorage.setItem).toHaveBeenCalled();
//   expect(AsyncStorage.setItem).toBeCalledWith('monthlyAverage');
//   expect(AsyncStorage.setItem).toBeCalledWith('monthlyComposition');
//   expect(AsyncStorage.setItem).toBeCalledWith('weeklyAverage');
//   expect(AsyncStorage.setItem).toBeCalledWith('weeklyComposition');
// });

// test('foodprint screen loads from cache when no internet connection', async () => {

// })
