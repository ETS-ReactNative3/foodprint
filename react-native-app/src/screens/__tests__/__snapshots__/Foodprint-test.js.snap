// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`foodprint screen with loaded results matches previous snapshot 1`] = `
<MockedProvider
  addTypename={false}
  mocks={
    Array [
      Object {
        "request": Object {
          "query": undefined,
        },
        "result": Object {
          "data": Object {
            "getUserHistoryReport": Object {
              "categoryReports": Array [
                Object {
                  "eggsAndDairy": Array [
                    Object {
                      "avgCarbonFootprint": 3.8,
                      "periodNumber": 0,
                    },
                    Object {
                      "avgCarbonFootprint": 3.3,
                      "periodNumber": -1,
                    },
                    Object {
                      "avgCarbonFootprint": 2.85,
                      "periodNumber": -2,
                    },
                    Object {
                      "avgCarbonFootprint": 2.6,
                      "periodNumber": -3,
                    },
                    Object {
                      "avgCarbonFootprint": 2.7,
                      "periodNumber": -4,
                    },
                    Object {
                      "avgCarbonFootprint": 2.6,
                      "periodNumber": -5,
                    },
                  ],
                  "fish": Array [
                    Object {
                      "avgCarbonFootprint": 3.8,
                      "periodNumber": 0,
                    },
                    Object {
                      "avgCarbonFootprint": 2.6,
                      "periodNumber": -1,
                    },
                    Object {
                      "avgCarbonFootprint": 3.3,
                      "periodNumber": -2,
                    },
                    Object {
                      "avgCarbonFootprint": 2.7,
                      "periodNumber": -3,
                    },
                    Object {
                      "avgCarbonFootprint": 2.85,
                      "periodNumber": -4,
                    },
                    Object {
                      "avgCarbonFootprint": 2.6,
                      "periodNumber": -5,
                    },
                  ],
                  "meat": Array [
                    Object {
                      "avgCarbonFootprint": 3.8,
                      "periodNumber": 0,
                    },
                    Object {
                      "avgCarbonFootprint": 3.3,
                      "periodNumber": -1,
                    },
                    Object {
                      "avgCarbonFootprint": 2.85,
                      "periodNumber": -2,
                    },
                    Object {
                      "avgCarbonFootprint": 2.6,
                      "periodNumber": -3,
                    },
                    Object {
                      "avgCarbonFootprint": 2.7,
                      "periodNumber": -4,
                    },
                    Object {
                      "avgCarbonFootprint": 3.3,
                      "periodNumber": -5,
                    },
                  ],
                  "plantBased": Array [
                    Object {
                      "avgCarbonFootprint": 3.8,
                      "periodNumber": 0,
                    },
                    Object {
                      "avgCarbonFootprint": 2.7,
                      "periodNumber": -1,
                    },
                    Object {
                      "avgCarbonFootprint": 3.3,
                      "periodNumber": -2,
                    },
                    Object {
                      "avgCarbonFootprint": 2.85,
                      "periodNumber": -3,
                    },
                    Object {
                      "avgCarbonFootprint": 2.6,
                      "periodNumber": -4,
                    },
                    Object {
                      "avgCarbonFootprint": 3.3,
                      "periodNumber": -5,
                    },
                  ],
                },
                Object {
                  "eggsAndDairy": Array [
                    Object {
                      "avgCarbonFootprint": 13.5,
                      "periodNumber": 0,
                    },
                    Object {
                      "avgCarbonFootprint": 16.5,
                      "periodNumber": -1,
                    },
                    Object {
                      "avgCarbonFootprint": 19,
                      "periodNumber": -2,
                    },
                    Object {
                      "avgCarbonFootprint": 13,
                      "periodNumber": -3,
                    },
                    Object {
                      "avgCarbonFootprint": 14.25,
                      "periodNumber": -4,
                    },
                    Object {
                      "avgCarbonFootprint": 13,
                      "periodNumber": -5,
                    },
                  ],
                  "fish": Array [
                    Object {
                      "avgCarbonFootprint": 13.5,
                      "periodNumber": 0,
                    },
                    Object {
                      "avgCarbonFootprint": 13,
                      "periodNumber": -1,
                    },
                    Object {
                      "avgCarbonFootprint": 16.5,
                      "periodNumber": -2,
                    },
                    Object {
                      "avgCarbonFootprint": 19,
                      "periodNumber": -3,
                    },
                    Object {
                      "avgCarbonFootprint": 14.25,
                      "periodNumber": -4,
                    },
                    Object {
                      "avgCarbonFootprint": 13,
                      "periodNumber": -5,
                    },
                  ],
                  "meat": Array [
                    Object {
                      "avgCarbonFootprint": 13.5,
                      "periodNumber": 0,
                    },
                    Object {
                      "avgCarbonFootprint": 16.5,
                      "periodNumber": -1,
                    },
                    Object {
                      "avgCarbonFootprint": 19,
                      "periodNumber": -2,
                    },
                    Object {
                      "avgCarbonFootprint": 13,
                      "periodNumber": -3,
                    },
                    Object {
                      "avgCarbonFootprint": 14.25,
                      "periodNumber": -4,
                    },
                    Object {
                      "avgCarbonFootprint": 16.5,
                      "periodNumber": -5,
                    },
                  ],
                  "plantBased": Array [
                    Object {
                      "avgCarbonFootprint": 13.5,
                      "periodNumber": 0,
                    },
                    Object {
                      "avgCarbonFootprint": 19,
                      "periodNumber": -1,
                    },
                    Object {
                      "avgCarbonFootprint": 16.5,
                      "periodNumber": -2,
                    },
                    Object {
                      "avgCarbonFootprint": 14.25,
                      "periodNumber": -3,
                    },
                    Object {
                      "avgCarbonFootprint": 13,
                      "periodNumber": -4,
                    },
                    Object {
                      "avgCarbonFootprint": 16.5,
                      "periodNumber": -5,
                    },
                  ],
                },
              ],
              "periodAvgs": Array [
                13.4,
                64,
              ],
              "userAvg": 17.25,
            },
          },
        },
      },
    ]
  }
>
  <Foodprint
    navigation={
      Object {
        "setOptions": [MockFunction],
      }
    }
  />
</MockedProvider>
`;
